heartbeat/monitors/active/tcp/tcp.go

```diff
@@ -39,113 +43,219 @@ func init() {
 
 var debugf = logp.MakeDebug("tcp")
 
-type connURL struct {
-	Scheme string
-	Host   string
-	Ports  []uint16
-}
-
 func create(
 	name string,
 	cfg *common.Config,
 ) (jobs []jobs.Job, endpoints int, err error) {
-	config := DefaultConfig
-	if err := cfg.Unpack(&config); err != nil {
+	return createWithResolver(cfg, monitors.StdResolver)
+}
+
+// Custom resolver is useful for tests against hostnames locally where we don't want to depend on any
+// hostnames existing in test environments
+func createWithResolver(
+	cfg *common.Config,
+	resolver monitors.Resolver,
+) (jobs []jobs.Job, endpoints int, err error) {
+	jc, err := makeJobFactory(cfg, resolver)
+	if err != nil {
 		return nil, 0, err
 	}
 
-	tls, err := tlscommon.LoadTLSConfig(config.TLS)
+	jobs, err = jc.makeJobs()
 	if err != nil {
 		return nil, 0, err
 	}
 
-	defaultScheme := "tcp"
-	if tls != nil {
-		defaultScheme = "ssl"
+	return jobs, len(jc.endpoints), nil
+}
+
+// jobFactory is where most of the logic here lives. It provides a common context around
+// the complex logic of executing a TCP check.
+type jobFactory struct {
+	config        Config
+	tlsConfig     *tlscommon.TLSConfig
+	defaultScheme string
+	endpoints     []endpoint
+	dataCheck     dataCheck
+	resolver      monitors.Resolver
+}
+
+func makeJobFactory(commonCfg *common.Config, resolver monitors.Resolver) (jf *jobFactory, err error) {
+	jf = &jobFactory{config: DefaultConfig, resolver: resolver}
+	err = jf.loadConfig(commonCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return jf, nil
+}
+
+// loadConfig parses the YAML config and populates the jobFactory fields.
+func (jf *jobFactory) loadConfig(commonCfg *common.Config) (err error) {
```

Comment:
same here, just: (error)

Reply:
Later in the function there's an assignment direct to a struct `jf.tlsConfig, err = tlscommon.LoadTLSConfig(jf.config.TLS)` member that requires the declaration of `err`. Using `:=` does not solve that. It's a weird thing in go

Reply:
Maybe I wasn't clear - I see a value in having a consistency in method definitions. Either name returned variables or not. If you insist on naming them (line 94), adjust all the places where there isn't a plain "return" command.

Personally I prefer not to name them and just define returned types. I'm fine with naming them, but not using a single "return" command is redundant.

Reply:
Consistency is very important! The style across heartbeat tends to be, if you need to initialize a var that's returned later, initialize it in the function return type definition.

Reply:
Correct, but adjust then actual `return` statements:

```go
func (jf *jobFactory) loadConfig(commonCfg *common.Config) (err error) {
	if err = commonCfg.Unpack(&jf.config); err != nil {
		return err
	}
```

```return err``` -> ```return```

Reply:
> Later in the function there's an assignment direct to a struct jf.tlsConfig, err = 

this can be solved by having a `var err error` before the call to tlsConfig, or store the result of LoadTLSConfig in a temporary variable. The `:=` is similar to `var`, but `jf.tlsConfig` is now new variable.

For Named returns I like the answer here: https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

> Don't name result parameters just to avoid declaring a var inside the function; that trades off a minor implementation brevity at the cost of unnecessary API verbosity

Reply:
Changed per the go styleguide